tests/st/harness/core/test_runner.py

```
@@ -232,185 +251,264 @@ def _compile_for_cache(test_case: "PTOTestCase", work_dir: Path, dump_passes: bo
+    waiting on slow compiles — workers only receive work that is ready to
+    run, so the device pool stays evenly loaded even when compile latency
+    varies between cases.
+    """
+    name = tc.get_name()
+
+    def _on_compile_done(cf: "Future") -> None:
         try:
-            _compile_for_cache(tc, work_dir, dump_passes)
-            return key, work_dir, None
+            artifact = cf.result()
         except Exception as exc:
```

Ruff (0.15.10)

[warning] Do not catch blind exception: `Exception`

(BLE001)